### ISSUE TITLE
GetPageAt;ExtractTextTo: When the number of PDF pages increases linearly, the cost of both methods increases exponentially

### ISSUE THREAD

**Issue by tayei1997** [podofo/podofo#149]
hello，

When I used podofo 0.10.2's `ExtractTextTo` function to extract the following three PDF files created by Ms Word, I found that the extraction time increased exponentially as the number of pages increased linearly:
[t20.pdf](https://github.com/podofo/podofo/files/14783537/t20.pdf)
[t1.pdf](https://github.com/podofo/podofo/files/14783535/t1.pdf)
[t10.pdf](https://github.com/podofo/podofo/files/14783536/t10.pdf)
t1. pdf is a 21 page PDF file that costs 2s to extract;
t1 Copy 10 copies of the PDF to get t10. pdf. Pdf, it takes about 70s to extract;
t1 Copy the contents of the PDF 20 copies to get t20. PDF, it takes about 300s to extract;

I further apply the following code to t20.pdf was tested and found that even within the same file, the larger the page number, the more time was consumed by `GetPageAt`, `ExtractTextTo`.  The code runs in VS2019 Release x64.
   ```
    TimerBase tb; 
    tb.Start();
    int skp = 63; //By only extracting 21 times the page number, to ensure that the page number of each extraction is the same
    for (unsigned i = 0; i < pages.GetCount(); i++)
    {
        if (skp * i >= pages.GetCount())
        {
            break;
        }
        auto et11 = tb.PrintElapsedMs();
        auto& page = pages.GetPageAt(skp * i);
        auto et22 = tb.PrintElapsedMs();
        std::cout << "page：" << skp * i << ",GetPageAt cost: " << (et22 - et11) / 1000.0 << "s" << std::endl;

        vector<PdfTextEntry> entries;

        auto et1 = tb.PrintElapsedMs();
        page.ExtractTextTo(entries);
        auto et2 = tb.PrintElapsedMs();
        std::cout <<"page："<< skp *i << ", ExtractTextTo cost: " << (et2 - et1)/1000.0 << "s" << std::endl;
        std::cout <<  std::endl;
    }
```
The results of the extraction are as follows:
![image](https://github.com/podofo/podofo/assets/72743757/58ba2cde-2019-433f-9cd4-8932cd5488dd)

Looking at the time-consuming difference  between `ExtractTextTo `and `GetPageAt` on each page, you can see that it is basically the same, so presumably there is a similar operation in both, assuming P, where p takes up most of the parsing time, and the cost increases with the page number, and the difference is the real page resolution time.
![image](https://github.com/podofo/podofo/assets/72743757/77b0f928-1801-43f1-9120-3cce20935abe)

Looking at the source code for `GetPageAt`  , it looks like the time it took to get each page should be similar.

**Comment by tayei1997** [comment]
I tried to extract a 765-page pdf file created by Adobe Acrobat Pro DC 20.6.20034, and it didn't seem to have the problem.
![image](https://github.com/podofo/podofo/assets/72743757/05416449-39e9-4788-a409-0f725f0f906a)

Compare the page structure of the two files
ms word  page structure:
![image](https://github.com/podofo/podofo/assets/72743757/cc65e8ff-d994-4b53-8090-846432ed9e20)
Adobe Acrobat  page structure:
![image](https://github.com/podofo/podofo/assets/72743757/eba30ea2-55a9-4860-90cc-cecbb4b450e9)

All of Word's pages are placed in the kids array, while Adobe's pages use pages as an intermediate node, organizing the pages into a tree-like structure.

**Comment by ceztko** [comment]
https://github.com/podofo/podofo/releases/tag/0.10.3

> Fixed big performance regression introduced in 0.10, see https://github.com/podofo/podofo/issues/108

Please always check changelogs of last official release. There may be other performance issues around but please note for the future that screenshots of PDF structures don't help me as much as as the output of a [profiler](https://github.com/podofo/podofo/issues/85#issuecomment-1595570370).

**Comment by tayei1997** [comment]
Thanks for your patience,
 I actually read the log and [#108](https://github.com/podofo/podofo/issues/108), but I mistakenly thought this was a performance issue related to resolution and rendering and metadata, not related to what I found.
I just tested it with 0.10.3 and the above problem has been resolved.
I will confirm before asking questions later. Thank you again for your patience.

**Comment by ceztko** [comment]
Actually there was a quite nasty performance issue (not a regression) that was fixed in [0.10.1](https://github.com/podofo/podofo/releases/tag/0.10.1):

> Rewritten PdfPageCollection for performance, see https://github.com/podofo/podofo/issues/85

That looked more similar to your issue. But the most relevant problem was fixed and discovered in 0.10.3: it was more pervasive as the more objects in the document, the more all operations were slowed down. That was a regression that was caused from augmenting the internal data structures: a general algorithms that was used for searches turned to perform quadratically in these conditions.
